### ISSUE TITLE
ElasticSearch: aggregation exception error

### ISSUE THREAD

**Issue by frknasir** [openownership/register#55]
# The Error

[400] {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Text fields are not optimised for operations that require per-document field data like aggregations and sorting, so these operations are disabled by default. Please use a keyword field instead. Alternatively, set fielddata=true on [type] in order to load field data by uninverting the inverted index. Note that this can use significant memory."}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"open_ownership_register_entities_development","node":"SLgzHYLrTfidSNSMn3-UeQ","reason":{"type":"illegal_argument_exception","reason":"Text fields are not optimised for operations that require per-document field data like aggregations and sorting, so these operations are disabled by default. Please use a keyword field instead. Alternatively, set fielddata=true on [type] in order to load field data by uninverting the inverted index. Note that this can use significant memory."}}],"caused_by":{"type":"illegal_argument_exception","reason":"Text fields are not optimised for operations that require per-document field data like aggregations and sorting, so these operations are disabled by default. Please use a keyword field instead. Alternatively, set fielddata=true on [type] in order to load field data by uninverting the inverted index. Note that this can use significant memory.","caused_by":{"type":"illegal_argument_exception","reason":"Text fields are not optimised for operations that require per-document field data like aggregations and sorting, so these operations are disabled by default. Please use a keyword field instead. Alternatively, set fielddata=true on [type] in order to load field data by uninverting the inverted index. Note that this can use significant memory."}}},"status":400}

# Stack Trace

elasticsearch-transport (6.1.0) lib/elasticsearch/transport/transport/base.rb:205:in `__raise_transport_error'
elasticsearch-transport (6.1.0) lib/elasticsearch/transport/transport/base.rb:323:in `perform_request'
elasticsearch-transport (6.1.0) lib/elasticsearch/transport/transport/http/faraday.rb:20:in `perform_request'
elasticsearch-transport (6.1.0) lib/elasticsearch/transport/client.rb:131:in `perform_request'
elasticsearch-api (6.1.0) lib/elasticsearch/api/actions/search.rb:187:in `search'
elasticsearch-model (6.0.0) lib/elasticsearch/model/searching.rb:51:in `execute!'
elasticsearch-model (6.0.0) lib/elasticsearch/model/response.rb:29:in `response'
elasticsearch-model (6.0.0) lib/elasticsearch/model/response/base.rb:34:in `total'
app/controllers/searches_controller.rb:12:in `show'
actionpack (5.2.4.3) lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
actionpack (5.2.4.3) lib/abstract_controller/base.rb:194:in `process_action'
actionpack (5.2.4.3) lib/action_controller/metal/rendering.rb:30:in `process_action'
actionpack (5.2.4.3) lib/abstract_controller/callbacks.rb:42:in `block in process_action'
activesupport (5.2.4.3) lib/active_support/callbacks.rb:132:in `run_callbacks'
actionpack (5.2.4.3) lib/abstract_controller/callbacks.rb:41:in `process_action'
actionpack (5.2.4.3) lib/action_controller/metal/rescue.rb:22:in `process_action'
actionpack (5.2.4.3) lib/action_controller/metal/instrumentation.rb:34:in `block in process_action'
activesupport (5.2.4.3) lib/active_support/notifications.rb:168:in `block in instrument'
activesupport (5.2.4.3) lib/active_support/notifications/instrumenter.rb:23:in `instrument'
activesupport (5.2.4.3) lib/active_support/notifications.rb:168:in `instrument'
actionpack (5.2.4.3) lib/action_controller/metal/instrumentation.rb:32:in `process_action'
actionpack (5.2.4.3) lib/action_controller/metal/params_wrapper.rb:256:in `process_action'
actionpack (5.2.4.3) lib/abstract_controller/base.rb:134:in `process'
actionview (5.2.4.3) lib/action_view/rendering.rb:32:in `process'
actionpack (5.2.4.3) lib/action_controller/metal.rb:191:in `dispatch'
actionpack (5.2.4.3) lib/action_controller/metal.rb:252:in `dispatch'
actionpack (5.2.4.3) lib/action_dispatch/routing/route_set.rb:52:in `dispatch'
actionpack (5.2.4.3) lib/action_dispatch/routing/route_set.rb:34:in `serve'
actionpack (5.2.4.3) lib/action_dispatch/journey/router.rb:52:in `block in serve'
actionpack (5.2.4.3) lib/action_dispatch/journey/router.rb:35:in `each'
actionpack (5.2.4.3) lib/action_dispatch/journey/router.rb:35:in `serve'
actionpack (5.2.4.3) lib/action_dispatch/routing/route_set.rb:840:in `call'
rack-attack (6.2.1) lib/rack/attack.rb:156:in `call'
rack-attack (6.2.1) lib/rack/attack.rb:170:in `call'
warden (1.2.8) lib/warden/manager.rb:36:in `block in call'
warden (1.2.8) lib/warden/manager.rb:34:in `catch'
warden (1.2.8) lib/warden/manager.rb:34:in `call'
rack (2.2.3) lib/rack/tempfile_reaper.rb:15:in `call'
rack (2.2.3) lib/rack/etag.rb:27:in `call'
rack (2.2.3) lib/rack/conditional_get.rb:27:in `call'
rack (2.2.3) lib/rack/head.rb:12:in `call'
actionpack (5.2.4.3) lib/action_dispatch/http/content_security_policy.rb:18:in `call'
rack (2.2.3) lib/rack/session/abstract/id.rb:266:in `context'
rack (2.2.3) lib/rack/session/abstract/id.rb:260:in `call'
actionpack (5.2.4.3) lib/action_dispatch/middleware/cookies.rb:670:in `call'
actionpack (5.2.4.3) lib/action_dispatch/middleware/callbacks.rb:28:in `block in call'
activesupport (5.2.4.3) lib/active_support/callbacks.rb:98:in `run_callbacks'
actionpack (5.2.4.3) lib/action_dispatch/middleware/callbacks.rb:26:in `call'
actionpack (5.2.4.3) lib/action_dispatch/middleware/executor.rb:14:in `call'
rollbar (2.27.0) lib/rollbar/middleware/rails/rollbar.rb:25:in `block in call'
rollbar (2.27.0) lib/rollbar.rb:145:in `scoped'
rollbar (2.27.0) lib/rollbar/middleware/rails/rollbar.rb:22:in `call'
actionpack (5.2.4.3) lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
rollbar (2.27.0) lib/rollbar/middleware/rails/show_exceptions.rb:22:in `call_with_rollbar'
web-console (3.4.0) lib/web_console/middleware.rb:135:in `call_app'
web-console (3.4.0) lib/web_console/middleware.rb:28:in `block in call'
web-console (3.4.0) lib/web_console/middleware.rb:18:in `catch'
web-console (3.4.0) lib/web_console/middleware.rb:18:in `call'
actionpack (5.2.4.3) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
railties (5.2.4.3) lib/rails/rack/logger.rb:38:in `call_app'
railties (5.2.4.3) lib/rails/rack/logger.rb:26:in `block in call'
activesupport (5.2.4.3) lib/active_support/tagged_logging.rb:71:in `block in tagged'
activesupport (5.2.4.3) lib/active_support/tagged_logging.rb:28:in `tagged'
activesupport (5.2.4.3) lib/active_support/tagged_logging.rb:71:in `tagged'
railties (5.2.4.3) lib/rails/rack/logger.rb:26:in `call'
sprockets-rails (3.2.1) lib/sprockets/rails/quiet_assets.rb:13:in `call'
actionpack (5.2.4.3) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
request_store (1.4.0) lib/request_store/middleware.rb:19:in `call'
actionpack (5.2.4.3) lib/action_dispatch/middleware/request_id.rb:27:in `call'
rack (2.2.3) lib/rack/method_override.rb:24:in `call'
rack (2.2.3) lib/rack/runtime.rb:22:in `call'
activesupport (5.2.4.3) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
actionpack (5.2.4.3) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (5.2.4.3) lib/action_dispatch/middleware/static.rb:127:in `call'
rack (2.2.3) lib/rack/sendfile.rb:110:in `call'
webpacker (4.0.7) lib/webpacker/dev_server_proxy.rb:29:in `perform_request'
rack-proxy (0.6.5) lib/rack/proxy.rb:57:in `call'
railties (5.2.4.3) lib/rails/engine.rb:524:in `call'
puma (4.3.5) lib/puma/configuration.rb:228:in `call'
puma (4.3.5) lib/puma/server.rb:713:in `handle_request'
puma (4.3.5) lib/puma/server.rb:472:in `process_client'
puma (4.3.5) lib/puma/server.rb:328:in `block in run'
puma (4.3.5) lib/puma/thread_pool.rb:134:in `block in spawn_thread'

# Solution

Changed the following code in `search.rb`:

```ruby
def self.aggregations
    {
      type: {
        terms: {
          field: :type
        },
      },
      country: {
        terms: {
          field: :country_code
        },
      },
    }
  end
```

to:

```ruby
def self.aggregations
    {
      type: {
        terms: {
          field: "#{:type}.keyword",
        },
      },
      country: {
        terms: {
          field: "#{:country_code}.keyword",
        },
      },
    }
  end
```

# Question

How is it that it worked for you as it is and not for me in my local environment. Do you have any pointers for me?

**Comment by stevenday** [comment]
Hi @frknasir,

That is odd, as those fields should be not be text fields - they're mapped to keywords in [app/models/entity.rb](https://github.com/openownership/register/blob/a3fbeea7f810c10d52ef688cb06367e84c45a405/app/models/entity.rb#L56-L62).

My immediate suggestion (assuming you don't have tons of data) would be to delete and recreate your search indexes:
```shell
bundle exec rails c
Entity.__elasticsearch__.create_index! force: true
Entity.import
```

and see if that resolves the problem.

On a separate note, is this for a personal project or something else? We'd love to know what you're using our code for if you can share - feel free to email us: tech@openownership.org.

**Comment by frknasir** [comment]
Hello @stevenday 

Thank you for the quick response. Now im getting this

```bash
Elasticsearch::Transport::Transport::Errors::BadRequest ([400] {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"The mapping definition cannot be nested under a type [_doc] unless include_type_name is set to true."}],"type":"illegal_argument_exception","reason":"The mapping definition cannot be nested under a type [_doc] unless include_type_name is set to true."},"status":400})
```

Yeah, so an organisation here in Nigeria is discussing the possible implementation of an open ownership register. I am helping them test out your open source code for possible usage. btw, I have not used ruby prior to this, let alone rails. I read up on them over the weekend and now i am here. There will, definitely, be an email from the organisation when they decide to move forward with it.

**Comment by frknasir** [comment]
It turned out to be compatibility between es 7 and elasticsearch-model 6.0. solved it by updating to elasticsearch-model 7.0 cc https://github.com/elastic/elasticsearch-rails/issues/877

Now, I am getting this:

<img width="1014" alt="Screenshot 2020-11-09 at 17 56 39" src="https://user-images.githubusercontent.com/4984175/98571572-f4b6a400-22b4-11eb-82a0-53f0a1559a07.png">

**Comment by frknasir** [comment]
> It turned out to be compatibility between es 7 and elasticsearch-model 6.0. solved it by updating to elasticsearch-model 7.0 cc [elastic/elasticsearch-rails#877](https://github.com/elastic/elasticsearch-rails/issues/877)
> 
> Now, I am getting this:
> 
> <img alt="Screenshot 2020-11-09 at 17 56 39" width="1014" src="https://user-images.githubusercontent.com/4984175/98571572-f4b6a400-22b4-11eb-82a0-53f0a1559a07.png">
> 
> perhaps, maybe es 7 should have been downgraded, instead of updating elasticsearch-model?

So this patch: https://github.com/elastic/elasticsearch-rails/issues/876#issuecomment-483773824 did the trick for me. Thank you!

**Comment by stevenday** [comment]
> an organisation here in Nigeria is discussing the possible implementation of an open ownership register. I am helping them test out your open source code for possible usage. btw, I have not used ruby prior to this, let alone rails. I read up on them over the weekend and now i am here. There will, definitely, be an email from the organisation when they decide to move forward with it.

Great! I know we've given some high-level support to Nigeria regarding their register, so we'd be really interested in understanding more about this project. 

> maybe es 7 should have been downgraded, instead of updating elasticsearch-model?

Ah, yes, this is your issues - we are only using (and testing with) version 6.x of Elasticsearch at the moment. If you use the included docker-compose.yml to run your local services via Docker, you'll get the exact same versions we use. The README has more on this: https://github.com/openownership/register#setting-up-local-development but probably doesn't emphasise enough that we really suggest using docker :)

Since I think this issue is solved one way or another, I'll close this now. For future support enquiries it would be better to contact us directly on tech@openownership.org as I don't always monitor these issues that closely.
